crates/cargo-util/src/paths.rs

```diff
@@ -190,14 +190,22 @@ pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()>
 /// Writes a file to disk atomically.
 ///
 /// This uses `tempfile::persist` to accomplish atomic writes.
+/// If the path is a symlink, it will follow the symlink and write to the actual target.
 pub fn write_atomic<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()> {
     let path = path.as_ref();
 
+    // Check if the path is a symlink and follow it if it is
+    let actual_path = if path.is_symlink() {
+        std::fs::read_link(path)?
+    } else {
+        path.to_path_buf()
+    };
```

Comment:
To avoid unnecessary allocations. And we can continue call it `path` so we don't need to change other lines.
(It is good that shadowing is allowed in Rust)

```suggestion
    let path = if path.is_symlink() {
        &std::fs::read_link(path)?
    } else {
        path
    };
```